File: android/src/main/java/com/pichillilorenzo/flutter_inappwebview/types/WebMessageListener.java
package com.pichillilorenzo.flutter_inappwebview.types;

import android.net.Uri;
import android.text.TextUtils;
import android.webkit.ValueCallback;
import android.webkit.WebView;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.webkit.JavaScriptReplyProxy;
import androidx.webkit.WebMessageCompat;
import androidx.webkit.WebViewCompat;
import androidx.webkit.WebViewFeature;

import com.pichillilorenzo.flutter_inappwebview.Util;
import com.pichillilorenzo.flutter_inappwebview.in_app_webview.InAppWebView;
import com.pichillilorenzo.flutter_inappwebview.plugin_scripts_js.JavaScriptBridgeJS;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;

import io.flutter.plugin.common.BinaryMessenger;
import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;

public class WebMessageListener implements MethodChannel.MethodCallHandler {
  static final String LOG_TAG = "WebMessageListener";

  public String jsObjectName;
  public Set<String> allowedOriginRules;
  public WebViewCompat.WebMessageListener listener;
  public JavaScriptReplyProxy replyProxy;
  public MethodChannel channel;
  public InAppWebViewInterface webView;

  public WebMessageListener(@NonNull InAppWebViewInterface webView, @NonNull BinaryMessenger messenger, @NonNull String jsObjectName, @NonNull Set<String> allowedOriginRules) {
    this.webView = webView;
    this.jsObjectName = jsObjectName;
    this.allowedOriginRules = allowedOriginRules;
    this.channel = new MethodChannel(messenger, "com.pichillilorenzo/flutter_inappwebview_web_message_listener_" + this.jsObjectName);
    this.channel.setMethodCallHandler(this);
    if (this.webView instanceof InAppWebView) {
      final WebMessageListener self = this;
      this.listener = new WebViewCompat.WebMessageListener() {
        @Override
        public void onPostMessage(@NonNull WebView view, @NonNull WebMessageCompat message, @NonNull Uri sourceOrigin, boolean isMainFrame, @NonNull JavaScriptReplyProxy javaScriptReplyProxy) {
          replyProxy = javaScriptReplyProxy;
          self.onPostMessage(message.getData(), sourceOrigin, isMainFrame);
        }
      };
    }
  }

  public void initJsInstance() {
    if (webView != null) {
      String jsObjectNameEscaped = Util.replaceAll(jsObjectName, "\'", "\\'");
      List<String> allowedOriginRulesStringList = new ArrayList<>();
      for (String allowedOriginRule : allowedOriginRules) {
        if ("*".equals(allowedOriginRule)) {
          allowedOriginRulesStringList.add("'*'");
        } else {
          Uri rule = Uri.parse(allowedOriginRule);
          String host = rule.getHost() != null ? "'" + Util.replaceAll(rule.getHost(), "\'", "\\'") + "'" : "null";
          allowedOriginRulesStringList.add("{scheme: '" + rule.getScheme() + "', host: " + host + ", port: " + (rule.getPort() != -1 ? rule.getPort() : "null") + "}");
        }
      }
      String allowedOriginRulesString = TextUtils.join(", ", allowedOriginRulesStringList);

      String source = "(function() {" +
              "  var allowedOriginRules = [" + allowedOriginRulesString + "];" +
              "  var isPageBlank = window.location.href === 'about:blank';" +
              "  var scheme = !isPageBlank ? window.location.protocol.replace(':', '') : null;" +
              "  var host = !isPageBlank ? window.location.hostname : null;" +
              "  var port = !isPageBlank ? window.location.port : null;" +
              "  if (window." + JavaScriptBridgeJS.JAVASCRIPT_BRIDGE_NAME + "._isOriginAllowed(allowedOriginRules, scheme, host, port)) {" +
              "      window['" + jsObjectNameEscaped + "'] = new FlutterInAppWebViewWebMessageListener('" + jsObjectNameEscaped + "');" +
              "  }" +
              "})();";
      webView.getUserContentController().addPluginScript(new PluginScript(
              "WebMessageListener-" + jsObjectName,
              source,
              UserScriptInjectionTime.AT_DOCUMENT_START,
              null,
              false
      ));
    }
  }

  @Nullable
  public static WebMessageListener fromMap(@NonNull InAppWebViewInterface webView, @NonNull BinaryMessenger messenger, @Nullable Map<String, Object> map) {
    if (map == null) {
      return null;
    }
    String jsObjectName = (String) map.get("jsObjectName");
    assert jsObjectName != null;
    List<String> allowedOriginRuleList = (List<String>) map.get("allowedOriginRules");
    assert allowedOriginRuleList != null;
    Set<String> allowedOriginRules = new HashSet<>(allowedOriginRuleList);
    return new WebMessageListener(webView, messenger, jsObjectName, allowedOriginRules);
  }

  @Override
  public void onMethodCall(@NonNull MethodCall call, @NonNull MethodChannel.Result result) {
    switch (call.method) {
      case "postMessage":
        if (webView instanceof InAppWebView) {
          String message = (String) call.argument("message");
          postMessageForInAppWebView(message, result);
        } else {
          result.success(true);
        }
        break;
      default:
        result.notImplemented();
    }
  }

  public void assertOriginRulesValid() throws Exception {
    int index = 0;
    for (String originRule : allowedOriginRules) {
      if (originRule == null) {
        throw new Exception("allowedOriginRules[" + index + "] is null");
      }
      if (originRule.isEmpty()) {
        throw new Exception("allowedOriginRules[" + index + "] is empty");
      }
      if ("*".equals(originRule)) {
        continue;
      }
      Uri url = Uri.parse(originRule);
      String scheme = url.getScheme();
      String host = url.getHost();
      String path = url.getPath();
      int port = url.getPort();
      if (scheme == null) {
        throw new Exception("allowedOriginRules " + originRule + " is invalid");
      }
      if (("http".equals(scheme) || "https".equals(scheme)) && (host == null || host.isEmpty())) {
        throw new Exception("allowedOriginRules " + originRule + " is invalid");
      }
      if (!"http".equals(scheme) && !"https".equals(scheme) && (host != null || port != -1)) {
        throw new Exception("allowedOriginRules " + originRule + " is invalid");
      }
      if ((host == null || host.isEmpty()) && port != -1) {
        throw new Exception("allowedOriginRules " + originRule + " is invalid");
      }
      if (!path.isEmpty()) {
        throw new Exception("allowedOriginRules " + originRule + " is invalid");
      }
      if (host != null) {
        int distance = host.indexOf("*");
        if (distance != 0 || (distance == 0 && !host.startsWith("*."))) {
          throw new Exception("allowedOriginRules " + originRule + " is invalid");
        }
        if (host.startsWith("[")) {
          if (!host.endsWith("]")) {
            throw new Exception("allowedOriginRules " + originRule + " is invalid");
          }
          String ipv6 = host.substring(1, host.length() - 1);
          if (!Util.isIPv6(ipv6)) {
            throw new Exception("allowedOriginRules " + originRule + " is invalid");
          }
        }
      }
      index++;
    }
  }

  private void postMessageForInAppWebView(String message, @NonNull MethodChannel.Result result) {
    if (replyProxy != null && WebViewFeature.isFeatureSupported(WebViewFeature.WEB_MESSAGE_LISTENER)) {
      replyProxy.postMessage(message);
    }
    result.success(true);
  }

  public boolean isOriginAllowed(String scheme, String host, int port) {
    for (String allowedOriginRule : allowedOriginRules) {
      if ("*".equals(allowedOriginRule)) {
        return true;
      }
      if (scheme == null || scheme.isEmpty()) {
        continue;
      }
      if ((scheme == null || scheme.isEmpty()) && (host == null || host.isEmpty()) && (port == 0 || port == -1)) {
        continue;
      }
      Uri rule = Uri.parse(allowedOriginRule);
      int rulePort = rule.getPort() == -1 || rule.getPort() == 0 ? ("https".equals(rule.getScheme()) ? 443 : 80) : rule.getPort();
      int currentPort = port == 0 || port == -1 ? ("https".equals(scheme) ? 443 : 80) : port;
      String IPv6 = null;
      if (rule.getHost() != null && rule.getHost().startsWith("[")) {
        try {
          IPv6 = Util.normalizeIPv6(rule.getHost().substring(1, rule.getHost().length() - 1));
        } catch (Exception ignored) {}
      }
      String hostIPv6 = null;
      try {
        hostIPv6 = Util.normalizeIPv6(host);
      } catch (Exception ignored) {}

      boolean schemeAllowed = rule.getScheme().equals(scheme);

      boolean hostAllowed = rule.getHost() == null ||
              rule.getHost().isEmpty() ||
              rule.getHost().equals(host) ||
              (rule.getHost().startsWith("*") && host != null && host.contains(rule.getHost().split("\\*")[1])) ||
              (hostIPv6 != null && IPv6 != null && hostIPv6.equals(IPv6));

      boolean portAllowed = rulePort == currentPort;

      if (schemeAllowed && hostAllowed && portAllowed) {
        return true;
      }
    }
    return false;
  }
  
  public void onPostMessage(String message, Uri sourceOrigin, boolean isMainFrame) {
    Map<String, Object> obj = new HashMap<>();
    obj.put("message", message);
    obj.put("sourceOrigin", sourceOrigin.toString().equals("null") ? null : sourceOrigin.toString());
    obj.put("isMainFrame", isMainFrame);
    channel.invokeMethod("onPostMessage", obj);
  }

  public void dispose() {
    this.channel.setMethodCallHandler(null);
    this.listener = null;
    this.replyProxy = null;
    this.webView = null;
  }
}

File: flutter_inappwebview_ios/ios/flutter_inappwebview_ios/Sources/flutter_inappwebview_ios/InAppWebView/WebMessage/WebMessageListener.swift
//
//  WebMessageListener.swift
//  flutter_inappwebview
//
//  Created by Lorenzo Pichilli on 10/03/21.
//

import Foundation
import WebKit
import Flutter

public class WebMessageListener: FlutterMethodCallDelegate {
    static var METHOD_CHANNEL_NAME_PREFIX = "com.pichillilorenzo/flutter_inappwebview_web_message_listener_"
    var id: String
    var jsObjectName: String
    var allowedOriginRules: Set<String>
    var channelDelegate: WebMessageListenerChannelDelegate?
    weak var webView: InAppWebView?
    var plugin: InAppWebViewFlutterPlugin?
    
    public init(plugin: InAppWebViewFlutterPlugin, id: String, jsObjectName: String, allowedOriginRules: Set<String>) {
        self.id = id
        self.plugin = plugin
        self.jsObjectName = jsObjectName
        self.allowedOriginRules = allowedOriginRules
        super.init()
        let channel = FlutterMethodChannel(name: WebMessageListener.METHOD_CHANNEL_NAME_PREFIX + self.id + "_" + self.jsObjectName,
                                           binaryMessenger: plugin.registrar.messenger())
        self.channelDelegate = WebMessageListenerChannelDelegate(webMessageListener: self, channel: channel)
    }
    
    public func assertOriginRulesValid() throws {
        for (index, originRule) in allowedOriginRules.enumerated() {
            if originRule.isEmpty {
                throw NSError(domain: "allowedOriginRules[\(index)] is empty", code: 0)
            }
            if originRule == "*" {
                continue
            }
            if let url = URL(string: originRule) {
                guard let scheme = url.scheme else {
                    throw NSError(domain: "allowedOriginRules \(originRule) is invalid", code: 0)
                }
                if scheme == "http" || scheme == "https", url.host == nil || url.host!.isEmpty {
                    throw NSError(domain: "allowedOriginRules \(originRule) is invalid", code: 0)
                }
                if scheme != "http", scheme != "https", url.host != nil || url.port != nil {
                    throw NSError(domain: "allowedOriginRules \(originRule) is invalid", code: 0)
                }
                if url.host == nil || url.host!.isEmpty, url.port != nil {
                    throw NSError(domain: "allowedOriginRules \(originRule) is invalid", code: 0)
                }
                if !url.path.isEmpty {
                    throw NSError(domain: "allowedOriginRules \(originRule) is invalid", code: 0)
                }
                if let hostname = url.host {
                    if let firstIndex = hostname.firstIndex(of: "*") {
                        let distance = hostname.distance(from: hostname.startIndex, to: firstIndex)
                        if distance != 0 || (distance == 0 && hostname.prefix(2) != "*.") {
                            throw NSError(domain: "allowedOriginRules \(originRule) is invalid", code: 0)
                        }
                    }
                    if hostname.hasPrefix("[") {
                        if !hostname.hasSuffix("]") {
                            throw NSError(domain: "allowedOriginRules \(originRule) is invalid", code: 0)
                        }
                        let fromIndex = hostname.index(hostname.startIndex, offsetBy: 1)
                        let toIndex = hostname.index(hostname.startIndex, offsetBy: hostname.count - 1)
                        let indexRange = Range<String.Index>(uncheckedBounds: (lower: fromIndex, upper: toIndex))
                        let ipv6 = String(hostname[indexRange])
                        if !Util.isIPv6(address: ipv6) {
                            throw NSError(domain: "allowedOriginRules \(originRule) is invalid", code: 0)
                        }
                    }
                }
            } else {
                throw NSError(domain: "allowedOriginRules \(originRule) is invalid", code: 0)
            }
        }
    }
    
    public func initJsInstance(webView: InAppWebView) {
        self.webView = webView
        if let webView = self.webView {
            let jsObjectNameEscaped = jsObjectName.replacingOccurrences(of: "\'", with: "\\'")
            let allowedOriginRulesString = allowedOriginRules.map { (allowedOriginRule) -> String in
                if allowedOriginRule == "*" {
                    return "'*'"
                }
                let rule = URL(string: allowedOriginRule)!
                let host = rule.host != nil ? "'" + rule.host!.replacingOccurrences(of: "\'", with: "\\'") + "'" : "null"
                return """
                {scheme: '\(rule.scheme!)', host: \(host), port: \(rule.port != nil ? String(rule.port!) : "null")}
                """
            }.joined(separator: ", ")
            let source = """
            (function() {
                \(WebMessageListener.isOriginAllowedJs)
            
                var allowedOriginRules = [\(allowedOriginRulesString)];
                var isPageBlank = window.location.href === "about:blank";
                var scheme = !isPageBlank ? window.location.protocol.replace(":", "") : null;
                var host = !isPageBlank ? window.location.hostname : null;
                var port = !isPageBlank ? window.location.port : null;
                if (_isOriginAllowed(allowedOriginRules, scheme, host, port)) {
                    window['\(jsObjectNameEscaped)'] = new FlutterInAppWebViewWebMessageListener('\(jsObjectNameEscaped)');
                }
            })();
            """
            
            let allowedOriginRules = webView.settings?.pluginScriptsOriginAllowList
            let forMainFrameOnly = webView.settings?.pluginScriptsForMainFrameOnly ?? true
            
            webView.configuration.userContentController.addPluginScript(PluginScript(
                groupName: "WebMessageListener-" + id + "-" + jsObjectName,
                source: source,
                injectionTime: .atDocumentStart,
                forMainFrameOnly: forMainFrameOnly,
                allowedOriginRules: allowedOriginRules,
                requiredInAllContentWorlds: false,
                messageHandlerNames: []
            ))
            webView.configuration.userContentController.sync(scriptMessageHandler: webView)
        }
    }
    
    public static func fromMap(plugin: InAppWebViewFlutterPlugin, map: [String:Any?]?) -> WebMessageListener? {
        guard let map = map else {
            return nil
        }
        return WebMessageListener(
            plugin: plugin,
            id: map["id"] as! String,
            jsObjectName: map["jsObjectName"] as! String,
            allowedOriginRules: Set(map["allowedOriginRules"] as! [String])
        )
    }
    
    public func isOriginAllowed(scheme: String?, host: String?, port: Int?) -> Bool {
        for allowedOriginRule in allowedOriginRules {
            if allowedOriginRule == "*" {
                return true
            }
            if scheme == nil || scheme!.isEmpty {
                continue
            }
            if scheme == nil || scheme!.isEmpty, host == nil || host!.isEmpty, port == nil || port == 0 {
                continue
            }
            if let rule = URL(string: allowedOriginRule) {
                let rulePort = rule.port == nil || rule.port == 0 ? (rule.scheme == "https" ? 443 : 80) : rule.port!
                let currentPort = port == nil || port == 0 ? (scheme == "https" ? 443 : 80) : port!
                var IPv6: String? = nil
                if let hostname = rule.host, hostname.hasPrefix("[") {
                    let fromIndex = hostname.index(hostname.startIndex, offsetBy: 1)
                    let toIndex = hostname.index(hostname.startIndex, offsetBy: hostname.count - 1)
                    let indexRange = Range<String.Index>(uncheckedBounds: (lower: fromIndex, upper: toIndex))
                    do {
                        IPv6 = try Util.normalizeIPv6(address: String(hostname[indexRange]))
                    } catch {}
                }
                var hostIPv6: String? = nil
                if let host = host, Util.isIPv6(address: host) {
                    do {
                        hostIPv6 = try Util.normalizeIPv6(address: host)
                    } catch {}
                }
                
                let schemeAllowed = scheme != nil && !scheme!.isEmpty && scheme == rule.scheme
                
                let hostAllowed = rule.host == nil ||
                    rule.host!.isEmpty ||
                    host == rule.host ||
                    (rule.host!.hasPrefix("*") && host != nil && host!.hasSuffix(rule.host!.split(separator: "*", omittingEmptySubsequences: false)[1])) ||
                    (hostIPv6 != nil && IPv6 != nil && hostIPv6 == IPv6)
                
                let portAllowed = rulePort == currentPort
                
                if schemeAllowed, hostAllowed, portAllowed {
                    return true
                }
            }
        }
        return false
    }
    
    private static let isOriginAllowedJs = """
        var _normalizeIPv6 = function(ip_string) {
            // replace ipv4 address if any
            var ipv4 = ip_string.match(/(.*:)([0-9]+\\.[0-9]+\\.[0-9]+\\.[0-9]+$)/);
            if (ipv4) {
                ip_string = ipv4[1];
                ipv4 = ipv4[2].match(/[0-9]+/g);
                for (var i = 0;i < 4;i ++) {
                    var byte = parseInt(ipv4[i],10);
                    ipv4[i] = ("0" + byte.toString(16)).substr(-2);
                }
                ip_string += ipv4[0] + ipv4[1] + ':' + ipv4[2] + ipv4[3];
            }
        
            // take care of leading and trailing ::
            ip_string = ip_string.replace(/^:|:$/g, '');
        
            var ipv6 = ip_string.split(':');
        
            for (var i = 0; i < ipv6.length; i ++) {
                var hex = ipv6[i];
                if (hex != "") {
                    // normalize leading zeros
                    ipv6[i] = ("0000" + hex).substr(-4);
                }
                else {
                    // normalize grouped zeros ::
                    hex = [];
                    for (var j = ipv6.length; j <= 8; j ++) {
                        hex.push('0000');
                    }
                    ipv6[i] = hex.join(':');
                }
            }
        
            return ipv6.join(':');
        };
        
        var _isOriginAllowed = function(allowedOriginRules, scheme, host, port) {
            for (var rule of allowedOriginRules) {
                if (rule === "*") {
                    return true;
                }
                if (scheme == null || scheme === "") {
                    continue;
                }
                if ((scheme == null || scheme === "") && (host == null || host === "") && (port === 0 || port === "" || port == null)) {
                    continue;
                }
                var rulePort = rule.port == null || rule.port === 0 ? (rule.scheme == "https" ? 443 : 80) : rule.port;
                var currentPort = port === 0 || port === "" || port == null ? (scheme == "https" ? 443 : 80) : port;
                var IPv6 = null;
                if (rule.host != null && rule.host[0] === "[") {
                    try {
                        IPv6 = _normalizeIPv6(rule.host.substring(1, rule.host.length - 1));
                    } catch {}
                }
                var hostIPv6 = null;
                try {
                    hostIPv6 = _normalizeIPv6(host);
                } catch {}
        
                var schemeAllowed = scheme == rule.scheme;
                
                var hostAllowed = rule.host == null ||
                    rule.host === "" ||
                    host === rule.host ||
                    (rule.host[0] === "*" && host != null && host.indexOf(rule.host.split("*")[1]) >= 0) ||
                    (hostIPv6 != null && IPv6 != null && hostIPv6 === IPv6);
        
                var portAllowed = rulePort === currentPort;
        
                if (schemeAllowed && hostAllowed && portAllowed) {
                    return true;
                }
            }
            return false;
        };
    """

    public func dispose() {
        channelDelegate?.dispose()
        channelDelegate = nil
        webView = nil
        plugin = nil
    }
    
    deinit {
        debugPrint("WebMessageListener - dealloc")
        dispose()
    }
}
